kolibri/plugins/coach/assets/src/modules/classSummary/dataHelpers.js

```diff
@@ -99,9 +99,17 @@ export default {
       }
       const groups = getters.getGroupIdsForAssignment(assignment);
       const adHocLearners = assignment.learner_ids;
-      return adHocLearners.length
-        ? uniq(adHocLearners.concat(getters.getLearnersForGroups(groups)))
-        : getters.getLearnersForGroups(groups);
+
+      let learners;
+      const totalLearnersInClass = Object.keys(state.learnerMap).length;
+
+      if (getters.getLearnersForGroups(groups).length === totalLearnersInClass) {
```

Comment:
I am very confused by this condition, it seems to be suggesting that if there are:
* groups where every learner in a class is assigned to one of those groups
* the assignment has been assigned to all of those groups
* and the assignment has also been assigned 'ad hoc' to a single learner

then we will only return the single learner as the learner for the assignment here?

I assume these changes were made to address the issue that @pcenov was seeing with the incorrect numbers in the reports and CSV export? Could you talk me through the logic here?

Reply:
> I am very confused by this condition, it seems to be suggesting that if there are:
> 
> * groups where every learner in a class is assigned to one of those groups
> * the assignment has been assigned to all of those groups
> * and the assignment has also been assigned 'ad hoc' to a single learner
> 
> then we will only return the single learner as the learner for the assignment here?
> 
> I assume these changes were made to address the issue that @pcenov was seeing with the incorrect numbers in the reports and CSV export? Could you talk me through the logic here?

Yes, so when I started with debugging this issue I found that [getters.getLearnersForGroups(groups)](https://github.com/learningequality/kolibri/blob/1f24cc880e88e40f19c8d4caa7934005b1a00721/kolibri/plugins/coach/assets/src/modules/classSummary/dataHelpers.js#L57) returns correct number of learners if a valid group is selected or else it defaults to the whole class. Now consider the case where we have only individual learners assigned then the following [code](https://github.com/learningequality/kolibri/blob/1f24cc880e88e40f19c8d4caa7934005b1a00721/kolibri/plugins/coach/assets/src/modules/classSummary/dataHelpers.js#L103) returns incorrect learner_ids i.e (whole class + ad-hoc Learners) but we wanted only the adhock_learners here. The implemented logic does the following thing:
1. It checks if the number of learners in the class is equivalent to the value returned by `getters.getLearnersForGroups(groups)` if this is the case we know that there are no valid groups available and the getter is returning the learner_ids of the whole class instead. So we then override the value of learners to just the adhock learners.
2. If that's not the case then we know that valid groups do exists and we would want to concatinate the adhock learners to it! i.e `uniq(adHocLearners.concat(getters.getLearnersForGroups(groups)));`.
3. And at last if the adhock learners are not there then we have assigned it to whole class and hence we return the default behaviour of `getters.getLearnersForGroups(groups)` i.e the learner_ids of whole class.

An alternate way that I thought off was to change the default behaviour of getter itself but considering code at other places might be dependent on it I proceeded with this solution.

Reply:
OK, so the condition in getLearnersForGroups is specifically when `groups` is an empty Array.

I would suggest we do the following instead:

```
const learnersFromGroups = groups.length ? getters.getLearnersForGroups(groups).length : [];
return uniq(adHocLearners.concat(learnersFromGroups));
```

This means we discard the `getLearnersFromGroups` in the case where we have no groups, which would give us the entire class of learners, but otherwise we preserve it. Then we unconditionally return the uniq set of all adhoc learners and those learners _only_ from the assigned groups.

Reply:
`const learnersFromGroups = groups.length ? getters.getLearnersForGroups(groups).length : [];`  this should be rather 
this right? `const learnersFromGroups = groups.length ? getters.getLearnersForGroups(groups): [];`

Reply:
Yes!